### ISSUE TITLE
Different module resolution results based on types: ['node']

### ISSUE THREAD

**Issue by gajus** [webpro/knip#317]
```ts
import * as RelationshipLayout from '@/components/Contact/ContactPage.styles.js';

const {
  EmptyStateActionsContainer,
  EmptyStateContainer,
  EmptyStateText,
  List,
} = RelationshipLayout;
```

Seems like Knip does not like this pattern. It reports all of these as unused.

**Comment by webpro** [comment]
Just extended some related fixture (also see the commit attached to this issue), but I can't reproduce this issue, is there more to it? Do you think it's a regression in the latest or a recent version?

Even if just destructuring the namespace (`RelationshipLayout`) already seems enough for them to be seen as used here.

**Comment by gajus** [comment]
I think I misidentified the issue.

This appears to be due to this pattern:

```ts
const CompleteOnboardingModal = lazy(async () => {
  return {
    default: (await import('@/components/CompleteOnboardingModal'))
      .CompleteOnboardingModal,
  };
});
```

`CompleteOnboardingModal` is being reported as unused, despite it clearly being used here.

**Comment by webpro** [comment]
I've tested and added a few more cases to the fixtures. See the attached commit.

Noticed that adding `"types": ["node"]` to the `tsconfig.json` helped TypeScript to resolve the modules properly.

Any chance you could share your `tsconfig.json`? Or is there other info I could look into.

**Comment by gajus** [comment]
> Noticed that adding "types": ["node"] to the tsconfig.json helped TypeScript to resolve the modules properly.

Ah, sorry missed that. Will try again and come back.

**Comment by webpro** [comment]
@ferferga created another issue as this looks unrelated to this issue.

**Comment by gajus** [comment]
I've some time debugging this this morning:

https://www.loom.com/share/ce594cd0bf21463c8fb18867736f9e32

The issue is def related to pattern:

```tsx
import * as ContactPageLayout from '@/components/Contact/ContactPage.styles.js';
const { AddButton, RoundButton } = ContactPageLayout;
```

Refactoring above to:

```tsx
import { AddButton, RoundButton } as ContactPageLayout from '@/components/Contact/ContactPage.styles.js';
```

Makes Knip detect that the component is used.

**Comment by gajus** [comment]
Wrote ESLint rule to prevent that pattern in our codebase:

https://github.com/gajus/eslint-plugin-canonical#no-import-namespace-destructure

**Comment by webpro** [comment]
Thanks for digging in, Gajus. While I appreciate the efforts, imho this type of destructuring should just work, are there other reasons to disallow it (other than an issue with Knip)?

Here is a minimal example of what I think has the scenarios of this issue: https://codesandbox.io/p/devbox/beautiful-dew-n2dq96

There is one unused export in the example (`B`). One of these changes results in zero unused exports:

- Add `"types": ["node"]`
- Install `typescript@latest` (v5.3)

I can't seem to reproduce the destructuring issue. Any chance you know what causes this?

**Comment by webpro** [comment]
The destructuring issue might be similar to the one reported in #353. _Maybe_ this fix in v2.43.0 helps you too?

**Comment by webpro** [comment]
I have a snapshot release with a potential fix:

```
npm install -D knip@node-types
```

This has `@types/node` + `types: ['node']` included with Knip.

Any chance you could try if this fixes the issue?

Extra bonus points if you could compare duration of a run against `knip@3.3.5` (you could use `time knip ...` or `knip --performance` for this).

(FYI there was a similar issue also fixed by this: #375).

**Comment by gajus** [comment]
Regarding timing:

```bash
# 3.3.0
$ knip  38.24s user 5.74s system 131% cpu 33.491 total
# 3.3.5
$ knip  38.29s user 5.81s system 131% cpu 33.647 total
```

**Comment by webpro** [comment]
> Regarding timing:
> 
> ```shell
> # 3.3.0
> $ knip  38.24s user 5.74s system 131% cpu 33.491 total
> # 3.3.5
> $ knip  38.29s user 5.81s system 131% cpu 33.647 total
> ```

Thanks! Looking good. Any chance you could add the `node-types` tagged one (`npm install -D knip@node-types`)? 🙏

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v3.4.0. See [Release 3.4.0](https://github.com/webpro/knip/releases/tag/3.4.0) for release notes._
